add support for BackedEnums to session

macbookandrew opened the pull request:
Similar to https://github.com/laravel/framework/pull/52677, https://github.com/laravel/framework/pull/52679, and https://github.com/laravel/framework/pull/52792, using enums improves code readability, maintainability and reduces the hassle of using string values within the code itself.

This adds BackedEnum support to the session methods:

```php
session()->put(RememberMeEnum::SESSION_KEY, true);
session()->forget(RememberMeEnum::SESSION_KEY, true);
```

This PR is backward compatible, with one minor(?) change to the `flash()` method: https://github.com/laravel/framework/pull/53235/files#r1806848088

This depends on https://github.com/laravel/framework/pull/53234 being merged first.

macbookandrew committed:
add support for BackedEnums to Arr helper

macbookandrew committed:
add support for backed enums to session store

macbookandrew committed:
clarify test values

macbookandrew commented on src/Illuminate/Session/Store.php:
```diff
@@ -449,11 +460,11 @@ public function decrement($key, $amount = 1)
      * @return void
      */
-    public function flash(string $key, $value = true)
+    public function flash($key, $value = true)
```
This is the one breaking change: removing the `string` type from the method.

I can revert this change and the doc if you like.

taylorotwell commented:
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If applicable, please consider releasing your code as a package so that the community can still take advantage of your contributions!

macbookandrew commented:
@taylorotwell I wonder if you would reconsider this one. I removed the changes to the `Arr::*` helper from #53234 so this focuses only on the session store, and updated to use `enum_value` instead of checking the type as requested in https://github.com/laravel/framework/pull/53234#discussion_r1807071249. 

The use case/pain point here is multiple Livewire components that are reading and writing values using the same session key; it would be really nice to use an enum for the session key instead of a string.

No matter what you decide, thanks for taking a look!
